test/common/http/header_map_impl_speed_test.cc

```diff
@@ -295,5 +295,65 @@ static void headerMapImplRemovePrefix(benchmark::State& state) {
 }
 BENCHMARK(headerMapImplRemovePrefix)->Arg(0)->Arg(1)->Arg(5)->Arg(10)->Arg(50);
 
+template <class HeaderMapType> class StaticLookupBenchmarker {
+public:
+  absl::optional<HeaderMapImpl::StaticLookupResponse> lookup(absl::string_view key) {
+    return table_.lookup(*ignored_, key);
+  }
+
+private:
+  std::unique_ptr<RequestHeaderMapImpl> ignored_ = RequestHeaderMapImpl::create();
+  HeaderMapImpl::StaticLookupTable<HeaderMapType> table_;
+};
+
+template <class HeaderMapType>
+static void headerMapImplStaticLookups(benchmark::State& state,
+                                       const std::vector<std::string>& keys) {
+  int i = keys.size();
+  StaticLookupBenchmarker<HeaderMapType> table;
+  for (auto _ : state) {
+    UNREFERENCED_PARAMETER(_);
+    auto result = table.lookup(keys[--i]);
+    if (i == 0) {
+      i = keys.size();
+    }
+    benchmark::DoNotOptimize(result);
+  }
+}
+
+static std::vector<std::string> makeMismatchedHeaders() {
+  return {
+      "x-envoy-banana",
+      "some-unknown-header",
+      "what-is-this-header",
+      "nobody-expects-this-header",
+      "another-unexpected-header",
+      "x-is-a-letter",
+      "x-y-problems-are-the-worst",
+  };
+}
+
+#define ADD_HEADER_TO_KEYS(name) keys.emplace_back(Http::Headers::get().name);
```

Comment:
nit: pass 'keys' to the macro rather than having this macro only work in a context where 'keys' is the local you want to modify

Reply:
Can't, because this macro is being passed to the other macros (`INLINE_REQ_HEADERS` and `INLINE_RESP_HEADERS`) as a macro-callback that only takes the one argument.

Reply:
I see; can you add a comment explaining it is to be used in that context?